packages/transaction-controller/src/TransactionController.ts

```diff
@@ -886,6 +896,57 @@ export class TransactionController extends BaseController<
     this.incomingTransactionHelper.stop();
   }
 
+  /**
+   * Adds external provided transaction to state as confirmed transaction.
+   *
+   * @param transactionMeta - TransactionMeta to add transactions.
+   * @param transactionReceipt - TransactionReceipt of the external transaction.
+   * @param baseFeePerGas - Base fee per gas of the external transaction.
+   */
+  async confirmExternalTransaction(
+    transactionMeta: TransactionMeta,
+    transactionReceipt: TransactionReceipt,
+    baseFeePerGas: Hex,
+  ) {
+    // Create logger
+    const log = createModuleLogger(projectLogger, 'confirmExternalTransaction');
+
+    // Run validation and add external transaction to state.
+    this.addExternalTransaction(transactionMeta);
+
+    try {
+      const transactionId = transactionMeta.id;
+      const gasUsed = normalizeTxReceiptGasUsed(
+        transactionReceipt.gasUsed as string,
+      );
+
+      // Make sure status is confirmed and define gasUsed as in receipt.
+      transactionMeta.status = TransactionStatus.confirmed;
+      transactionMeta.txReceipt = {
+        ...transactionReceipt,
+        gasUsed,
+      };
+      if (baseFeePerGas) {
+        transactionMeta.baseFeePerGas = baseFeePerGas;
+      }
+
+      // Update same nonce local transactions as dropped and define replacedBy properties.
+      this.markNonceDuplicatesDropped(transactionId);
+
+      // Update external provided transaction with updated gas values and confirmed status.
+      this.updateTransaction(transactionMeta);
+
+      // If provided transaction is a swap, then try updating txMeta.postTxBalance.
+      if (transactionMeta.type === TransactionType.swap) {
```

Comment:
Both `TransactionType` and Swaps arebig concepts that we already have dedicated tickets for so we can ensure we implement them fully and don't miss any usages.

Would it be safer and simpler to remove any type and swaps logic in this PR so we can split the work as much as possible and focus purely on the external confirm logic here without introducing other feature dependencies?

Reply:
Sure why not, I removed all swaps related types & private/util functions for now.
This is the removal commit for future reference: https://github.com/MetaMask/core/pull/1625/commits/5037c5d0a1fc385c9d3b2ab9449586e6e853bcb6